File: sort_examples/cocktailSort.js
sort = (array, film) => {
  let compares = 0;
  const greens = [];
  let start = 0;
  let end = array.length - 1;
  let swapped = true;

  while (start < end && swapped) {
    swapped = false;
    for (let i = start; i < end; i++) {
      const j = i + 1;
      const temp = array[j];
      film.rec({ array, i, j, temp, compares, greens });
      compares++;
      if (array[i] > array[j]) {
        array[j] = array[i];
        array[i] = temp;
        film.rec({ array, i, j, temp, compares, greens });
        swapped = true;
      }
    }
    greens.push(end);
    film.rec({ array, i: -1, j: -1, temp: array[end], compares, greens });
    end--;
    if (!swapped || start >= end) break;

    swapped = false;
    for (let i = end; i > start; i--) {
      const j = i - 1;
      const temp = array[j];
      film.rec({ array, i, j, temp, compares, greens });
      compares++;
      if (array[j] > array[i]) {
        array[j] = array[i];
        array[i] = temp;
        film.rec({ array, i, j, temp, compares, greens });
        swapped = true;
      }
    }
    greens.push(start);
    film.rec({ array, i: -1, j: -1, temp: array[start], compares, greens });
    start++;
  }

  if (array.length > 0) {
    for (let i = start; i <= end; i++) {
      greens.push(i);
    }
    const finalIndex = Math.min(start, array.length - 1);
    film.rec({
      array,
      i: -1,
      j: -1,
      temp: array[finalIndex],
      compares,
      greens,
    });
  }
};

File: sort_examples/cocktail_sort.py
def sort(array):
    start = 0
    end = len(array) - 1
    swapped = True

    while swapped:
        swapped = False
        for index in range(start, end):
            if array[index] > array[index + 1]:
                array.swap(index, index + 1)
                swapped = True

        array.mark_sorted(end)
        end -= 1
        if not swapped:
            break

        swapped = False
        for index in range(end, start, -1):
            if array[index - 1] > array[index]:
                array.swap(index - 1, index)
                swapped = True

        array.mark_sorted(start)
        start += 1
